filebeat/input/kafka/input.go

```diff
@@ -325,33 +283,162 @@ func (h *groupHandler) ack(message *sarama.ConsumerMessage) {
+	groupHandler *groupHandler
+	log          *logp.Logger
+}
+
+func (m *recordReader) Close() error {
+	return nil
+}
+
+func (m *recordReader) Next() (reader.Message, error) {
+	msg, ok := <-m.claim.Messages()
+	if !ok {
+		return reader.Message{}, io.EOF
+	}
+
+	timestamp, kafkaFields := composeEventMetadata(m.claim, m.groupHandler, msg)
+	ackHandler := func() {
+		m.groupHandler.ack(msg)
+	}
+	return composeMessage(timestamp, msg.Value, kafkaFields, ackHandler), nil
+}
+
+type listFromFieldReader struct {
+	claim        sarama.ConsumerGroupClaim
+	groupHandler *groupHandler
+	buffer       []reader.Message
+	field        string
+	log          *logp.Logger
+}
+
+func (l *listFromFieldReader) Close() error {
 	return nil
 }
 
+func (l *listFromFieldReader) Next() (reader.Message, error) {
+	if len(l.buffer) != 0 {
+		return l.returnFromBuffer()
+	}
+
+	msg, ok := <-l.claim.Messages()
+	if !ok {
+		return reader.Message{}, io.EOF
+	}
+
+	timestamp, kafkaFields := composeEventMetadata(l.claim, l.groupHandler, msg)
+	messages := parseMultipleMessages(msg.Value, l.field, l.log)
+
+	neededAcks := atomic.MakeInt(len(messages))
+	ackHandler := func() {
+		if neededAcks.Dec() == 0 {
+			l.groupHandler.ack(msg)
+		}
+	}
+	for _, message := range messages {
+		newBuffer := append(l.buffer, composeMessage(timestamp, []byte(message), kafkaFields, ackHandler))
+		l.buffer = newBuffer
+	}
+
+	return l.returnFromBuffer()
+}
+
+func (l *listFromFieldReader) returnFromBuffer() (reader.Message, error) {
+	next := l.buffer[0]
+	newBuffer := l.buffer[1:]
+	l.buffer = newBuffer
+	return next, nil
+}
+
+func composeEventMetadata(claim sarama.ConsumerGroupClaim, handler *groupHandler, msg *sarama.ConsumerMessage) (time.Time, common.MapStr) {
+	timestamp := time.Now()
+	kafkaFields := common.MapStr{
+		"topic":     claim.Topic(),
+		"partition": claim.Partition(),
+		"offset":    msg.Offset,
+		"key":       string(msg.Key),
+	}
+
+	version, versionOk := handler.version.Get()
+	if versionOk && version.IsAtLeast(sarama.V0_10_0_0) {
+		timestamp = msg.Timestamp
+		if !msg.BlockTimestamp.IsZero() {
+			kafkaFields["block_timestamp"] = msg.BlockTimestamp
+		}
+	}
+	if versionOk && version.IsAtLeast(sarama.V0_11_0_0) {
+		kafkaFields["headers"] = arrayForKafkaHeaders(msg.Headers)
+	}
+	return timestamp, kafkaFields
+}
+
+func composeMessage(timestamp time.Time, content []byte, kafkaFields common.MapStr, ackHandler func()) reader.Message {
+	return reader.Message{
+		Ts:      timestamp,
+		Content: content,
+		Fields: common.MapStr{
+			"kafka":   kafkaFields,
+			"message": string(content),
+		},
+		Meta: common.MapStr{
+			"ackHandler": ackHandler,
+		},
+	}
+}
+
 // parseMultipleMessages will try to split the message into multiple ones based on the group field provided by the configuration
-func (h *groupHandler) parseMultipleMessages(bMessage []byte) []string {
+func parseMultipleMessages(bMessage []byte, field string, log *logp.Logger) []string {
```

Comment:
maybe this can be a function of `listFromFieldReader`?

Reply:
ah yes I initially  left it unchanged from the original code but this is better, done.